src/Processors/QueryPipeline.cpp

```diff
@@ -255,6 +262,95 @@ QueryPipeline QueryPipeline::unitePipelines(
     return pipeline;
 }
 
+std::unique_ptr<QueryPipeline> QueryPipeline::joinPipelines(
+    std::unique_ptr<QueryPipeline> left,
+    std::unique_ptr<QueryPipeline> right,
+    JoinPtr join,
+    size_t max_block_size,
+    Processors * collected_processors)
+{
+    left->checkInitializedAndNotCompleted();
+    right->checkInitializedAndNotCompleted();
+
+    /// Extremes before join are useless. They will be calculated after if needed.
+    left->pipe.dropExtremes();
+    right->pipe.dropExtremes();
+
+    left->pipe.collected_processors = collected_processors;
+    right->pipe.collected_processors = collected_processors;
+
+    /// In case joined subquery has totals, and we don't, add default chunk to totals.
+    bool default_totals = false;
+    if (!left->hasTotals() && right->hasTotals())
+    {
+        left->addDefaultTotals();
+        default_totals = true;
+    }
+
+    ///                                     (left)   ──────> Joining ─> (joined)
+    ///                                     (left)   ─┐┌───>
+    ///                                   (totals)   ─┼┼─┐
+    ///                                               └┼─┼─> Joining ─> (joined)
+    ///                                                │┌┼─>
+    /// (right)  ─>                                  ──┘││
+    /// (right)  ─> Resize ─> AddingJoined ─> Resize ───┘└─> Joining ─> (totals)
+    /// (totals) ───────────>                        ──────>
```

Comment:
Nice diagram. Two entries for (right) and Joining mean parallel execution, right? I didn't realize right away where the splits/merges are, maybe it makes sense to picture them with `┤` and the like.

Reply:
> Two entries for (right) and Joining mean parallel execution, right?

Yes.

> maybe it makes sense to picture them with ┤ and the like

Good idea!

Reply:
I've spend like 20 minutes trying to make it better, but unicode seem to be not so powerful...